src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
+[`GenericArg`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/subst/struct.GenericArg.html)
+is essentially a space-efficient wrapper around
+[`GenericArgKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/subst/enum.GenericArgKind.html),
+which is an enum indicating what kind of generic the type parameter is (type, lifetime, or const).
+Thus, `SubstsRef` is conceptually like a `&'tcx [GenericArgKind<'tcx>]` slice (but it is actually a
+`List`).
+
+So why do we use this `List` type instead of making it really a slice? The reason is that we may
+want to compare subslices of the list of generics for equality. For example, imagine we had a `List`
+`a = [u32, f32]` and  a `List` `b = [f32]`. Now suppose that we have a subslice of `a` : `a2 =
+[f32]`. Recall that lists are interned, so if we want to compare `a` with `b`, we can just compare
+their pointers and see whether they are equivalent or not. But we may *also* want to check the
+equality of `b` and `a2` (i.e. compare subslices of lists). Since the subslice `a2`  would also be
+interned, it would actually end up being the same as `b`, so we would be able to efficiently compare
+them by just comparing the pointers (just like with the full lists). You wouldn’t be able to do that
+with normal slices and would have to iterate over the contents to check for equality.
+
+So pulling it all together, let’s go back to our example above:
+
+```rust,ignore
+struct MyStruct<T> { x: u32, y: T }
+```
+
+- There would be an `AdtDef` (and corresponding `DefId`) for `MyStruct`.
+- There would be a `TyKind::Param` (and corresponding `DefId`) for `T` (more later).
+- There would be a `SubstsRef` containing the list `[GenericArgKind::Type(Ty(T))]`
+    - The `Ty(T)` here is my shorthand for entire other `ty::Ty` that has `TyKind::Param`, which we
+      mentioned in the previous point.
+- This is one `TyKind::Adt` containing the `AdtDef` of `MyStruct` with the `SubstsRef` above.
+
+Finally, we will quickly mention the
+[`Generics`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.Generics.html) type. It
+is used to give information about the type parameters of a type.
+
+### Unsubstituted Generics
+
+So above, recall that in our example the `MyStruct` struct had a generic type `T`. When we are (for
+example) type checking functions that use `MyStruct`, we will need to be able to refer to this type
+`T` without actually knowing what it is. In general, this is true inside all generic definitions: we
+need to be able to work with unknown types. This is done via `TyKind::Param` (which we mentioned in
+the example above).
+
+Each `TyKind::Param` contains two things: the index of the type parameter and the name. The more
+important one is the index, since the name can change across scopes (e.g. we can define struct
+`MyStruct<A>` by then implement `impl<B> MyStruct<B>` ; the names are different, but they mean the
+same type parameter).
```

Comment:
This is actually rather incorrect -- they are not the same type parameter, but they happen to play the same role in this instance. Consider `impl<B> MyStruct<Vec<B>>` -- here, the index of `B` is zero, but it is not "the same" as `A`. In fact, `A` (in the definition) will be replaced with `Vec<B>`.

More generally, we should never mix code with type parameters from distinct scopes -- so you should never simultaneously have a a type referencing `B` (defined on the impl) with a type referencing `A` (defined on the struct). Instead, you would always substitute `A` for some other type.

I would say it more like this:

* It contains the name and the index. In general, the index fully defines the parameter and is used by most of the code. The name is included for debug print-outs. There are two reasons for this. First, the index is convenient, it allows you to include into the list of generic arguments when substituting. Second, the index is more robust. For example, you could in principle have two distinct type parameters that use the same name, e.g. `impl<A> Foo<A> { fn bar<A>() { .. } }`, although the rules against shadowing make this difficult (but those language rules could change in the future).

Reply:
Done.